dask/array/gufunc.py

```diff
@@ -0,0 +1,396 @@
+from __future__ import absolute_import, division, print_function
+
+from itertools import count
+from operator import getitem
+import re
+from numpy import vectorize as np_vectorize
+from numpy import broadcast_to as np_broadcast_to
+
+try:
+    from cytoolz import concat, merge
+except ImportError:
+    from toolz import concat, merge
+from functools import partial
+
+from .core import asarray, Array, atop, broadcast_shapes
+from ..core import flatten
+from .. import sharedict
+
+
+_valid_name_re = re.compile("^[_a-zA-Z][_a-zA-Z0-9]*$")
+
+
+def _parse_dim(dim):
+    if not dim:
+        return None
+    assert len(_valid_name_re.findall(dim)) == 1
+    return dim
+
+
+def _parse_arg(sig):
+    if not sig:
+        return None
+    assert len(re.findall("^\([_a-zA-Z0-9,]*\)$", sig)) == 1
+    sig = sig[1:-1]
+    dims = sig.split(',')
+    ret = [_parse_dim(d) for d in dims]
+    return tuple(r for r in ret if r is not None)
+
+
+def _parse_args(sig):
+    if not sig:
+        return None
+    assert len(re.findall("^(\([_a-zA-Z0-9,]*\)),?(,?\([_a-zA-Z0-9,]*\))*$", sig)) == 1
+    sigs = re.findall("\([_a-zA-Z0-9,]*\)*", sig)
+    ret = [_parse_arg(s) for s in sigs]
+    return list(r for r in ret if r is not None)
+
+
+def parse_signature(signature):
```

Comment:
You might just borrow the version of this function that I wrote for `numpy.vectorize`:
https://github.com/numpy/numpy/blob/99019107cbde06294a356fd7fc859f9b31f87410/numpy/lib/function_base.py#L1638-L1666